sycl/source/device_selector.cpp

```diff
@@ -157,5 +160,138 @@ int host_selector::operator()(const device &dev) const {
   return Score;
 }
 
+namespace detail {
+std::string trim_spaces(std::string input) {
+  size_t LStart = input.find_first_not_of(" ");
+  std::string LTrimmed =
+      (LStart == std::string::npos) ? "" : input.substr(LStart);
+
+  size_t REnd = LTrimmed.find_last_not_of(" ");
+  return (REnd == std::string::npos) ? "" : LTrimmed.substr(0, REnd + 1);
+}
+
+std::vector<std::string> tokenize(std::string filter, std::string delim) {
+  std::vector<std::string> Tokens;
+  size_t Pos = 0;
+  std::string Input = filter;
+  std::string Tok;
+
+  while ((Pos = Input.find(delim)) != std::string::npos) {
+    Tok = Input.substr(0, Pos);
+    Input.erase(0, Pos + delim.length());
+    // Erase leading and trailing WS
+    Tok = trim_spaces(Tok);
+
+    if (Tok.size() > 0)
+      Tokens.push_back(Tok);
+  }
+
+  if (Input.size() > 0)
+    Input = trim_spaces(Input);
+
+  // Add remainder
+  if (Input.size() > 0)
+    Tokens.push_back(Input);
+
+  return Tokens;
+}
+
+enum TokenType { TokPlatform, TokDeviceType, TokUnknown };
+
+TokenType parse_kind(std::string token) {
+  TokenType Result = TokUnknown;
+
+  if (token.find("platform") != std::string::npos)
+    Result = TokPlatform;
+  if (token.find("type") != std::string::npos)
+    Result = TokDeviceType;
+
+  return Result;
+}
+
+std::string strip_kind(std::string token) {
+  std::string Prefix = "=";
+  size_t Loc = token.find(Prefix);
+
+  if (Loc == std::string::npos)
+    return token;
+
+  // move past the '='
+  Loc = Loc + 1;
+
+  return token.substr(Loc);
+}
+
+bool match(std::string input, std::string pattern) {
+  return (input.find(pattern) != std::string::npos);
+}
+} // namespace detail
+
+namespace ext {
+namespace oneapi {
+string_selector::string_selector(std::string filter) {
+  std::transform(filter.begin(), filter.end(), filter.begin(),
+                 [](unsigned char c) { return std::tolower(c); });
+  std::vector<std::string> Tokens = detail::tokenize(filter, ";");
+
+  for (auto Tok : Tokens) {
+    if (Tok.find("=") == std::string::npos)
+      continue;
+
+    detail::TokenType TTy = detail::parse_kind(Tok);
+    std::string SubTok = detail::strip_kind(Tok);
+    std::vector<std::string> SubTokens = detail::tokenize(SubTok, ",");
+
+    if (TTy == detail::TokPlatform) {
+      mPlatforms.insert(mPlatforms.end(), SubTokens.begin(), SubTokens.end());
+    } else if (TTy == detail::TokDeviceType) {
+      mDeviceTypes.insert(mDeviceTypes.end(), SubTokens.begin(),
+                          SubTokens.end());
+    } else {
+      throw runtime_error("Invalid string_selector input! Please specify at "
+                          "least one platform or device type filter.",
+                          PI_INVALID_VALUE);
+    }
+  }
+}
+
+int string_selector::operator()(const device &dev) const {
+  int Score = REJECT_DEVICE_SCORE;
+
+  std::string CPU = "cpu";
+  std::string GPU = "gpu";
```

Comment:
Are there plans to support accelerator(s) eventually as well?

Reply:
Yes, this can be extended with more types in the future as necessary.